api-lib/sendSocialMessage.ts

```diff
@@ -0,0 +1,96 @@
+import fetch from 'node-fetch';
+
+import { gql } from './Gql';
+
+const DISCORD_BOT_NAME = 'Caesar The Coordinape Bot';
+
+type SocialMessage = {
+  message: string;
+  circleId: number;
+  sanitize: boolean;
+  channels: {
+    discord?: boolean;
+    telegram?: boolean;
+  };
+};
+
+function cleanStr(str: string) {
+  return str.replace(/:|-|\/|\*|_|`/g, '');
+}
+
+export async function sendSocialMessage({
+  message,
+  circleId,
+  sanitize = true,
+  channels,
+}: SocialMessage) {
+  const msg = sanitize ? cleanStr(message) : message;
+
+  const { circles } = await gql.q('query')({
+    circles: [
+      {
+        where: {
+          id: {
+            _eq: circleId,
+          },
+        },
+      },
+      {
+        discord_webhook: true,
+        telegram_id: true,
+      },
+    ],
+  });
+
+  const [{ discord_webhook, telegram_id }] = circles;
+
+  if (channels?.discord && discord_webhook) {
+    try {
+      const discordWebhookPost = {
+        content: msg,
+        username: DISCORD_BOT_NAME,
+        // avatar_url: '' TODO
+      };
+      const res = await fetch(discord_webhook, {
+        method: 'POST',
+        body: JSON.stringify(discordWebhookPost),
+        headers: {
+          'Content-Type': 'application/json',
+        },
+      });
+
+      if (!res.ok) {
+        throw new Error(JSON.stringify(await res.json()));
+      }
+    } catch (e) {
+      // TODO: what/where do we want to log?
```

Comment:
Would also be nice to be able to retry these for example if Discord goes down, how do we retry them once it's API is back up?

Reply:
Hasura has its internal ping logging system so we can just start with that. If we return the proper HTTP error responses in the event handler, we can configure retries etc within Hasura: https://hasura.io/docs/latest/graphql/core/event-triggers/create-trigger.html#retry-logic

In this case I would probably setup separate triggers / handlers for discord and telegram webhooks so that we can manage their error state and retries individually.

Reply:
If only `sendSocialMessage` throws, the mutation that updates `nominees.ended` as `true` has already been committed in the database. Thus the nominee that was updated on the first handler execution won't be picked up on the retry.

Is there a way to rollback a GraphQL mutation on error?

Reply:
Maybe you can try @META-DREAMER 's suggestion and use an event trigger onto nominee table's update event
then you won't need to do error handling since it runs async in the background

https://hasura.io/docs/latest/graphql/core/event-triggers/payload.html
could just within the event handler logic do a check on the payload old vs new ended =  false > true if yes trigger the send social message, and when creating the event trigger you can set the retries/timeout/intervals.

Reply:
Ah right, I didn't fully understand what I read - I'll take a look at event triggers next. One more thing: `check-nominee.ts` already queries nominees with this where clause:

```
          where: {
            ended: {
              _eq: false,
            },
            expiry_date: { _lte: new Date().toISOString() },
          },
```

If the event trigger handler should also check the payload for `nominee.ended = false > true` etc., isn't that essentially duplicating the business logic into two places?

Reply:
Yeah, I agree - I was just wondering if duplicating this sort of logic into to places could be confusing in the future if there's a lot of similar use cases (cron trigger + event handler), but there also doesn't seem to be a better option if we want to use Hasura retry logic. Anyhow, I'll work on the event triggers based on this convo 👍

Reply:
> https://hasura.io/docs/latest/graphql/core/event-triggers/payload.html could just within the event handler logic do a check on the payload old vs new nominee.ended = false > true + its past the nominee expiry date & insufficient vouches. if yes trigger the send social message, and when creating the event trigger you can set the retries/timeout/intervals.

Yes this is the ideal way to do it because it guarantees that a notification will be sent no matter how that field gets updated. e.g. If theres another mutation we create in the future also updates this field then we don't have to duplicate sending the notification. Unidirectional data flow as much as possible.